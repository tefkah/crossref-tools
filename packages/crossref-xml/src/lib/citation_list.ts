import { CitationList, Citation, Author, JournalTitle, ArticleTitle, Doi } from '../types'

export interface Cite {
  publishedWhere: string
  authors: [{ name: string }] | { name: string }[]
  title: string
  prefix: string
  suffix: string
  publishedAt: string | Date
}

export const citationList = ({ citations }: { citations: Cite[] }): CitationList => {
  const js: CitationList = {
    type: 'element',
    name: 'citation_list',
    children: citations.map((citation, index) => {
      const { publishedWhere, authors, title, prefix, suffix, publishedAt } = citation

      const datetime = typeof publishedAt === 'string' ? new Date(publishedAt) : publishedAt
      const citationJs: Citation = {
        type: 'element',
        name: 'citation',
        attributes: {
          // index starts at 0
          key: `${index + 1}`,
        },
        children: [
          {
            type: 'element',
            name: 'journal_title',
            children: [
              {
                type: 'text',
                value: publishedWhere,
              },
            ],
          } as JournalTitle,
          {
            type: 'element',
            name: 'author',
            children: [
              {
                type: 'text',
                value: authors[0].name,
              },
            ],
          } as Author,
          {
            type: 'element',
            name: 'cYear',
            children: [
              {
                type: 'text',
                value: datetime.getUTCFullYear().toString(),
              },
            ],
          },
          {
            type: 'element',
            name: 'doi',
            children: [
              {
                type: 'text',
                value: `${prefix}/${suffix}`,
              },
            ],
          } as Doi,
          // {
          //   type: "element",
          //   name: "isbn",
          //   children: [
          //     {
          //       type: "text",
          //       value: isbn,
          //     },
          //   ],
          // },
          {
            type: 'element',
            name: 'article_title',
            children: [
              {
                type: 'text',
                value: title,
              },
            ],
          } as ArticleTitle,
        ],
      }
      return citationJs
    }),
  }

  return js
}
