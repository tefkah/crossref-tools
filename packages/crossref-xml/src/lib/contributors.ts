// import { Element, Text } from "xast"

export interface Author {
  firstName?: string | null
  lastName?: string | null
  orcid?: string | null
}

import { u } from 'unist-builder'
import { x } from 'xastscript'
import { GivenName, Surname, PersonName, Contributors, ORCID } from '../types'

const dbAuthorToCrossrefAuthor = (author: Author): PersonName['children'] => {
  /**
   * Object.entries returns [key:string, value:any][], always
   * so we need to typecast it
   */
  return (
    (Object.entries(author) as [keyof Author, string][])
      .map(([key, value]) => {
        if (!value) {
          return null
        }

        switch (key) {
          case 'firstName':
            return x('given_name', [u('text', value)]) as GivenName
          case 'lastName':
            return x('surname', [u('text', value)]) as Surname

          case 'orcid':
            return x('ORCID', { authenticated: true }, [
              u('text', `https://orcid.org/${value}`),
            ]) as ORCID
        }
      })
      /**
       * Filter out null values
       */
      .filter((x) => x) as PersonName['children']
  )
}

// TODO: Feature update needed for organizations / affiliations
const contributors = (authors: Author[]): Contributors => {
  return x(
    'contributors',
    authors.map(
      (author, index) =>
        x(
          'person_name',
          {
            sequence: index === 0 ? 'first' : 'additional',
            contributor_role: 'author',
          },
          dbAuthorToCrossrefAuthor(author),
        ) as PersonName,
    ),
  ) as Contributors
}

export default contributors
