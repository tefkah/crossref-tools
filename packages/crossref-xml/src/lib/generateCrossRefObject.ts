import { Element, Instruction, Root } from 'xast'

import { head } from './head'
import { body } from './body'
import { Cite } from './citation_list'
import { Author } from './contributors'
import { URI } from './ai_program'
import { Head, Body } from '../types'

export interface CrossRef extends Root {
  children: (Instruction | DOIBatch)[]
}
export interface DOIBatch extends Element {
  name: 'doi_batch'
  attributes: {
    version: string
    xmlns: string
    'xmlns:xsi': string
    'xsi:schemaLocation': string
    'xmlns:jats': string
  }
  children: (Head | Body)[]
}
export const generateCrossRefObject = ({
  schema,
  type,
  title,
  language,
  authors,
  abstractText,
  license_url,
  doi,
  resolve_url,
  citations,
}: {
  schema: string
  type: string
  title: string
  language: string
  authors: Author[]
  abstractText: string
  license_url: URI
  doi: string
  resolve_url: URI
  citations: Cite[]
}): CrossRef => {
  const jsCrossRef: CrossRef = {
    type: 'root',
    children: [
      {
        type: 'instruction',
        name: 'xml',
        value: 'version="1.0" encoding="UTF-8"',
      },
      {
        type: 'element',
        name: 'doi_batch',
        attributes: {
          version: schema,
          xmlns: `http://www.crossref.org/schema/${schema}`,
          'xmlns:xsi': 'http://www.w3.org/2001/XMLSchema-instance',
          'xsi:schemaLocation': `http://www.crossref.org/schema/${schema} http://www.crossref.org/schemas/crossref${schema}.xsd`,
          'xmlns:jats': 'http://www.ncbi.nlm.nih.gov/JATS1',
        },
        children: [
          head({
            registrant: 'Liberate Science GmbH',
            email: 'info@libscie.org',
          }),
          body({
            type,
            title,
            language,
            authors,
            abstractText,
            license_url,
            doi,
            resolve_url,
            citations,
          }),
        ],
      },
    ],
  }

  return jsCrossRef
}
