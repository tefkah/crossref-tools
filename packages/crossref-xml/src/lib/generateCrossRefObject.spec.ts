import fs from 'fs'
import { Cite } from './citation_list'
import { Author } from './contributors'
import { generateCrossRefObject } from './generateCrossRefObject'

describe('generateCrossRefObject', () => {
  it('should generate a CrossRef object', () => {
    const schema = '3.0.4'
    const type = 'journal-article'
    const title = 'A title'
    const language = 'en'
    const authors: Author[] = [
      {
        firstName: 'John',
        lastName: 'Doe',
        orcid: '0000-0000-0000-0000',
      },
    ]
    const abstractText = 'An abstract'
    const license_url = 'https://creativecommons.org/licenses/by/4.0/'
    const doi = '10.1234/foo.bar'
    const resolve_url = 'https://doi.org/10.1234/foo.bar'
    const citations: Cite[] = [
      {
        prefix: '10.1234',
        suffix: 'bar',
        authors: [
          {
            name: 'Jane Doe',
          },
        ],
        title: 'A title',
        publishedAt: new Date().toUTCString(),
        publishedWhere: 'A place',
      },
    ]
    const crossRef = generateCrossRefObject({
      schema,
      type,
      title,
      language,
      authors,
      abstractText,
      license_url,
      doi,
      resolve_url,
      citations,
    })
    expect(crossRef).toMatchSnapshot()
  })
})
