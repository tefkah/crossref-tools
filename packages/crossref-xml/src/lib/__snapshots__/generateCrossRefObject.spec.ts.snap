// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`generateCrossRefObject should generate a CrossRef object 1`] = `
Object {
  "children": Array [
    Object {
      "name": "xml",
      "type": "instruction",
      "value": "version=\\"1.0\\" encoding=\\"UTF-8\\"",
    },
    Object {
      "attributes": Object {
        "version": "3.0.4",
        "xmlns": "http://www.crossref.org/schema/3.0.4",
        "xmlns:jats": "http://www.ncbi.nlm.nih.gov/JATS1",
        "xmlns:xsi": "http://www.w3.org/2001/XMLSchema-instance",
        "xsi:schemaLocation": "http://www.crossref.org/schema/3.0.4 http://www.crossref.org/schemas/crossref3.0.4.xsd",
      },
      "children": Array [
        Object {
          "attributes": Object {},
          "children": Array [
            Object {
              "attributes": Object {},
              "children": Array [
                Object {
                  "type": "text",
                  "value": "1661695672068",
                },
              ],
              "name": "doi_batch_id",
              "type": "element",
            },
            Object {
              "attributes": Object {},
              "children": Array [
                Object {
                  "type": "text",
                  "value": "1661695672068",
                },
              ],
              "name": "timestamp",
              "type": "element",
            },
            Object {
              "attributes": Object {},
              "children": Array [
                Object {
                  "attributes": Object {},
                  "children": Array [
                    Object {
                      "type": "text",
                      "value": "info@libscie.org",
                    },
                  ],
                  "name": "email_address",
                  "type": "element",
                },
              ],
              "name": "depositor",
              "type": "element",
            },
            Object {
              "type": "text",
              "value": "Liberate Science GmbH",
            },
          ],
          "name": "head",
          "type": "element",
        },
        Object {
          "children": Array [
            Object {
              "attributes": Object {
                "book_type": "other",
              },
              "children": Array [
                Object {
                  "attributes": Object {
                    "language": "en",
                  },
                  "children": Array [
                    Object {
                      "attributes": Object {},
                      "children": Array [
                        Object {
                          "attributes": Object {},
                          "children": Array [
                            Object {
                              "type": "text",
                              "value": "ResearchEquals",
                            },
                          ],
                          "name": "title",
                          "type": "element",
                        },
                      ],
                      "name": "titles",
                      "type": "element",
                    },
                    Object {
                      "attributes": Object {
                        "media_type": "online",
                      },
                      "children": Array [
                        Object {
                          "attributes": Object {},
                          "children": Array [
                            Object {
                              "type": "text",
                              "value": "08",
                            },
                          ],
                          "name": "month",
                          "type": "element",
                        },
                        Object {
                          "attributes": Object {},
                          "children": Array [
                            Object {
                              "type": "text",
                              "value": "28",
                            },
                          ],
                          "name": "day",
                          "type": "element",
                        },
                        Object {
                          "attributes": Object {},
                          "children": Array [
                            Object {
                              "type": "text",
                              "value": "2022",
                            },
                          ],
                          "name": "year",
                          "type": "element",
                        },
                      ],
                      "name": "publication_date",
                      "type": "element",
                    },
                    Object {
                      "attributes": Object {
                        "reason": "simple_series",
                      },
                      "children": Array [],
                      "name": "noisbn",
                      "type": "element",
                    },
                    Object {
                      "attributes": Object {},
                      "children": Array [
                        Object {
                          "attributes": Object {},
                          "children": Array [
                            Object {
                              "attributes": Object {},
                              "children": Array [],
                              "name": "text",
                              "type": "element",
                            },
                          ],
                          "name": "publisher_name",
                          "type": "element",
                        },
                        Object {
                          "attributes": Object {},
                          "children": Array [
                            Object {
                              "attributes": Object {},
                              "children": Array [],
                              "name": "text",
                              "type": "element",
                            },
                          ],
                          "name": "publisher_place",
                          "type": "element",
                        },
                      ],
                      "name": "publisher",
                      "type": "element",
                    },
                    Object {
                      "attributes": Object {
                        "name": "AccessIndicators",
                      },
                      "children": Array [
                        Object {
                          "children": Array [],
                          "name": "free_to_read",
                          "type": "element",
                        },
                        Object {
                          "attributes": Object {
                            "applies_to": "vor",
                          },
                          "children": Array [
                            Object {
                              "type": "text",
                              "value": "https://creativecommons.org/publicdomain/zero/1.0/legalcode",
                            },
                          ],
                          "name": "license_ref",
                          "type": "element",
                        },
                      ],
                      "name": "program",
                      "type": "element",
                    },
                    Object {
                      "attributes": Object {},
                      "children": Array [
                        Object {
                          "attributes": Object {},
                          "children": Array [
                            Object {
                              "type": "text",
                              "value": "10.53962/0000",
                            },
                          ],
                          "name": "doi",
                          "type": "element",
                        },
                        Object {
                          "attributes": Object {},
                          "children": Array [
                            Object {
                              "type": "text",
                              "value": "https://researchequals.com",
                            },
                          ],
                          "name": "resource",
                          "type": "element",
                        },
                      ],
                      "name": "doi_data",
                      "type": "element",
                    },
                  ],
                  "name": "book_metadata",
                  "type": "element",
                },
                Object {
                  "attributes": Object {
                    "component_type": "other",
                    "language": "en",
                  },
                  "children": Array [
                    Object {
                      "attributes": Object {},
                      "children": Array [
                        Object {
                          "attributes": Object {
                            "contributor_role": "author",
                            "sequence": "first",
                          },
                          "children": Array [
                            Object {
                              "attributes": Object {},
                              "children": Array [
                                Object {
                                  "type": "text",
                                  "value": "John",
                                },
                              ],
                              "name": "given_name",
                              "type": "element",
                            },
                            Object {
                              "attributes": Object {},
                              "children": Array [
                                Object {
                                  "type": "text",
                                  "value": "Doe",
                                },
                              ],
                              "name": "surname",
                              "type": "element",
                            },
                            Object {
                              "attributes": Object {
                                "authenticated": "true",
                              },
                              "children": Array [
                                Object {
                                  "type": "text",
                                  "value": "https://orcid.org/0000-0000-0000-0000",
                                },
                              ],
                              "name": "ORCID",
                              "type": "element",
                            },
                          ],
                          "name": "person_name",
                          "type": "element",
                        },
                      ],
                      "name": "contributors",
                      "type": "element",
                    },
                    Object {
                      "attributes": Object {},
                      "children": Array [
                        Object {
                          "attributes": Object {},
                          "children": Array [
                            Object {
                              "type": "text",
                              "value": "A title",
                            },
                          ],
                          "name": "title",
                          "type": "element",
                        },
                      ],
                      "name": "titles",
                      "type": "element",
                    },
                    Object {
                      "children": Array [
                        Object {
                          "children": Array [
                            Object {
                              "type": "text",
                              "value": "An abstract",
                            },
                          ],
                          "name": "jats:p",
                          "type": "element",
                        },
                      ],
                      "name": "jats:abstract",
                      "type": "element",
                    },
                    Object {
                      "attributes": Object {
                        "media_type": "online",
                      },
                      "children": Array [
                        Object {
                          "attributes": Object {},
                          "children": Array [
                            Object {
                              "type": "text",
                              "value": "08",
                            },
                          ],
                          "name": "month",
                          "type": "element",
                        },
                        Object {
                          "attributes": Object {},
                          "children": Array [
                            Object {
                              "type": "text",
                              "value": "28",
                            },
                          ],
                          "name": "day",
                          "type": "element",
                        },
                        Object {
                          "attributes": Object {},
                          "children": Array [
                            Object {
                              "type": "text",
                              "value": "2022",
                            },
                          ],
                          "name": "year",
                          "type": "element",
                        },
                      ],
                      "name": "publication_date",
                      "type": "element",
                    },
                    Object {
                      "attributes": Object {
                        "name": "AccessIndicators",
                      },
                      "children": Array [
                        Object {
                          "children": Array [],
                          "name": "free_to_read",
                          "type": "element",
                        },
                        Object {
                          "attributes": Object {
                            "applies_to": "vor",
                          },
                          "children": Array [
                            Object {
                              "type": "text",
                              "value": "https://creativecommons.org/licenses/by/4.0/",
                            },
                          ],
                          "name": "license_ref",
                          "type": "element",
                        },
                      ],
                      "name": "program",
                      "type": "element",
                    },
                    Object {
                      "attributes": Object {},
                      "children": Array [
                        Object {
                          "attributes": Object {},
                          "children": Array [
                            Object {
                              "type": "text",
                              "value": "10.1234/foo.bar",
                            },
                          ],
                          "name": "doi",
                          "type": "element",
                        },
                        Object {
                          "attributes": Object {},
                          "children": Array [
                            Object {
                              "type": "text",
                              "value": "https://doi.org/10.1234/foo.bar",
                            },
                          ],
                          "name": "resource",
                          "type": "element",
                        },
                      ],
                      "name": "doi_data",
                      "type": "element",
                    },
                    Object {
                      "children": Array [
                        Object {
                          "attributes": Object {
                            "key": "1",
                          },
                          "children": Array [
                            Object {
                              "children": Array [
                                Object {
                                  "type": "text",
                                  "value": "A place",
                                },
                              ],
                              "name": "journal_title",
                              "type": "element",
                            },
                            Object {
                              "children": Array [
                                Object {
                                  "type": "text",
                                  "value": "Jane Doe",
                                },
                              ],
                              "name": "author",
                              "type": "element",
                            },
                            Object {
                              "children": Array [
                                Object {
                                  "type": "text",
                                  "value": "2022",
                                },
                              ],
                              "name": "cYear",
                              "type": "element",
                            },
                            Object {
                              "children": Array [
                                Object {
                                  "type": "text",
                                  "value": "10.1234/bar",
                                },
                              ],
                              "name": "doi",
                              "type": "element",
                            },
                            Object {
                              "children": Array [
                                Object {
                                  "type": "text",
                                  "value": "A title",
                                },
                              ],
                              "name": "article_title",
                              "type": "element",
                            },
                          ],
                          "name": "citation",
                          "type": "element",
                        },
                      ],
                      "name": "citation_list",
                      "type": "element",
                    },
                    Object {
                      "attributes": Object {},
                      "children": Array [],
                      "name": "component_list",
                      "type": "element",
                    },
                  ],
                  "name": "content_item",
                  "type": "element",
                },
              ],
              "name": "book",
              "type": "element",
            },
          ],
          "name": "body",
          "type": "element",
        },
      ],
      "name": "doi_batch",
      "type": "element",
    },
  ],
  "type": "root",
}
`;
